pkg/authorization/apibinding_authorizer.go

```diff
@@ -0,0 +1,137 @@
+	}
+}
+
+type apiBindingAccessAuthorizer struct {
+	versionedInformers clientgoinformers.SharedInformerFactory
+	apiBindingIndexer  cache.Indexer
+	delegate           authorizer.Authorizer
+}
+
+func (a *apiBindingAccessAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	// Determine if resource being requested is bound.
+	// If it is bound get the workspace.
+	// request for the user in the cluster.
+
+	apiBindingAccessDenied := fmt.Sprintf("api %v.%v access is not permitted", attr.GetResource(), attr.GetAPIGroup())
+
+	// get the cluster from the ctx.
+	lcluster, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, apiBindingAccessDenied, err
+	}
+
+	bindingLogicalCluster, bound, err := a.getAPIBindingWorkspace(attr, lcluster)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, apiBindingAccessDenied, err
+	}
+
+	if !bound {
+		return a.delegate.Authorize(ctx, attr)
+	}
+
+	// If bound, create a rbac authorizer filtered to the cluster.
+	clusterKubeInformer := rbacwrapper.FilterInformers(bindingLogicalCluster, a.versionedInformers.Rbac().V1())
+	clusterAuthorizer := rbac.New(
+		&rbac.RoleGetter{Lister: clusterKubeInformer.Roles().Lister()},
+		&rbac.RoleBindingLister{Lister: clusterKubeInformer.RoleBindings().Lister()},
+		&rbac.ClusterRoleGetter{Lister: clusterKubeInformer.ClusterRoles().Lister()},
+		&rbac.ClusterRoleBindingLister{Lister: clusterKubeInformer.ClusterRoleBindings().Lister()},
+	)
+	dec, _, err := clusterAuthorizer.Authorize(ctx, attr)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, apiBindingAccessDenied, err
+	}
+
+	if dec == authorizer.DecisionAllow {
+		return a.delegate.Authorize(ctx, attr)
+	}
+
+	return authorizer.DecisionNoOpinion, apiBindingAccessDenied, nil
+}
+
+//TODO [shawn-hurley]: this should be a helper shared.
+func (a *apiBindingAccessAuthorizer) getAPIBindingWorkspace(attr authorizer.Attributes, clusterName logicalcluster.Name) (logicalcluster.Name, bool, error) {
+	parentClusterName, hasParent := clusterName.Parent()
+	if !hasParent {
+		// APIBindings in root are not possible (they can only point to sibling workspaces).
+		return logicalcluster.New(""), false, nil
+	}
+
+	objs, err := a.apiBindingIndexer.ByIndex(byWorkspaceIndex, clusterName.String())
+	if err != nil {
+		return logicalcluster.New(""), false, err
+	}
+	for _, obj := range objs {
+		apiBinding := obj.(*apisv1alpha1.APIBinding)
+		for _, br := range apiBinding.Status.BoundResources {
+			if apiBinding.Status.BoundAPIExport.Workspace == nil {
+				// this will never happen today. But as soon as we add other reference types (like exports), this log output will remind out of necessary work here.
+				continue
+			}
+			if br.Group == attr.GetAPIGroup() && br.Resource == attr.GetResource() {
+				return parentClusterName.Join(apiBinding.Status.BoundAPIExport.Workspace.WorkspaceName), true, nil
```

Comment:
@s-urbaniak @ncdc please build an opinion where we want to have the upper-bound policy. Here it is in the APIExport workspace. Is that good, or does it mangle the policy with the APIExport objects? Am not sure. Alternative: have a field in the APIExport for the policy workspace? (can all be follow-up).

Reply:
Added a field for this as discussed below